imap_processing/codice/codice_l2.py

```diff
@@ -13,24 +13,194 @@
+            L2_GEOMETRIC_FACTOR
+            * L2_HI_NUMBER_OF_SSD
+            * species_efficiencies
+            * energy_passbands
+        )
+        # Store by replacing existing species data with omni-directional intensities
+        l2_dataset[species].values = omni_direction_intensities
+
+    return l2_dataset
+
+
+def process_hi_sectored(
+    l2_dataset: xr.Dataset, dependencies: ProcessingInputCollection
+) -> xr.Dataset:
+    """
+    Process the hi-omni L1B dataset to calculate omni-directional intensities.
+
+    See section 11.1.2 of the CoDICE algorithm document for details.
+
+    The formula for omni-directional intensities is:
+        l1b species data / (geometric_factor * efficiency * energy_passband)
+
+    Geometric factor is constant for all species and is 0.013.
+    Efficiency is provided in a CSV file for each species and energy bin and
+    position.
+    Energy passband is calculated from energy_bin_minus + energy_bin_plus
+
+    Parameters
+    ----------
+    l2_dataset : xarray.Dataset
+        The L2 dataset to process.
+    dependencies : ProcessingInputCollection
+        The collection of processing input files.
+
+    Returns
+    -------
+    xarray.Dataset
+        The updated L2 dataset with omni-directional intensities calculated.
+    """
+    efficiencies_file = dependencies.get_file_paths(
+        descriptor="l2-hi-sectored-efficiency"
+    )[0]
+    efficiencies_df = pd.read_csv(efficiencies_file)
+    # Similar to hi-omni, each species has different shape.
+    # Because of that, we need to loop over each species and calculate
+    # sectored intensities separately.
+    for species in HI_SECTORED_VARIABLE_NAMES:
+        # Efficiencies from dataframe maps to different dimension in L1B data.
+        # For example:
+        #   l1b species 'h' has shape:
+        #       (epoch, 8, 12, 12) -> (time, energy, spin_sector, inst_az)
+        #   efficiencies 'h' has shape after reading from CSV:
+        #       (8, 12) -> (energy, inst_az)
+        #       NOTE: 12 here maps to last 12 in above l1b dimension.
+        # Because of this, it's easier to work with the data in xarray.
+        # Xarray automatically aligns dimensions and coordinates, making it easier
+        # to work with multi-dimensional data. Thus, we convert the efficiencies
+        # to xarray.DataArray with dimensions (energy, ssd_index)
+        # TODO: update ssd_index to inst_az when Joey data is updated.
+        species_data = efficiencies_df[efficiencies_df["species"] == species].values
+        species_efficiencies = xr.DataArray(
+            species_data[:, 2:].astype(
+                float
+            ),  # Skip first two columns (species, energy_bin)
+            dims=(f"energy_{species}", "ssd_index"),
+            coords={
+                f"energy_{species}": l2_dataset[f"energy_{species}"],
+                "ssd_index": l2_dataset["ssd_index"],
+            },
```

Comment:
Out of curiosity, can you simplify this by indexing into the original dataset itself? Maybe even removing the dims above if that gets carried along automatically too?
```suggestion
            coords=l2_dataset[[f"energy_{species}", "ssd_index"]]
            },
```

Reply:
yep. That works!